File: 백준/Gold/1300. K번째 수/K번째 수.py
import sys
input = sys.stdin.readline

N = int(input())
k = int(input())

start = 1
end = k
answer = 0

while start <= end:
    mid = (start + end) // 2
    total_count = 0

    for i in range(1, N + 1):
        row_count = mid // i
        if row_count >= N:
            row_count = N
        total_count += row_count

    if total_count < k:
        start = mid + 1
    else:
        end = mid - 1
        answer = mid
    
print(answer)

File: 백준/Gold/1300. K번째 수/K번째 수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.Math;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int k = Integer.parseInt(br.readLine());

        int start = 1;
        int end = k;

        int answer = 0;
        while (start <= end) {
            int mid = (start + end) / 2;

            int count = 0;
            for (int i = 1; i <= N; i++) {
                count += Math.min(mid / i, N);
            }

            if (count < k)
                start = mid + 1;
            else {
                answer = mid;
                end = mid - 1;
            }
        }

        System.out.println(answer);
    }
}
